src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -75,16 +93,250 @@ namespace
+        const char* typeAttrName = GetWellKnownAttributeName(WellKnownAttribute::UnsafeAccessorTypeAttribute);
+
+        // Determine the max parameter count. +1 for the return value, which is always index 0.
+        const uint32_t totalParamCount = cxt.DeclarationMetaSig.NumFixedArgs() + 1;
+
+        // Inspect all parameters to the declaration method for UnsafeAccessorTypeAttribute.
+        uint32_t attrCount = 0;
+        IMDInternalImport *pInternalImport = cxt.Declaration->GetModule()->GetMDImport();
+        HENUMInternalHolder hEnumParams(pInternalImport);
+        hEnumParams.EnumInit(mdtParamDef, cxt.Declaration->GetMemberDef());
+        mdParamDef currParamDef = mdParamDefNil;
+        while (hEnumParams.EnumNext(&currParamDef))
+        {
+            const void *pData;
+            ULONG cbData;
+            HRESULT hr = IfFailThrow(pInternalImport->GetCustomAttributeByName(currParamDef, typeAttrName, &pData, &cbData));
+            if (hr != S_OK)
+                continue;
+
+            // The first time we find an attribute, we allocate the translations array.
+            if (attrCount == 0)
+                cxt.TranslatedParams = new ParamDetails[totalParamCount];
+
+            // Parse the attribute data.
+            CustomAttributeParser cap(pData, cbData);
+            IfFailThrow(cap.ValidateProlog());
+            LPCUTF8 typeString;
+            ULONG typeStringLen;
+            IfFailThrow(cap.GetNonNullString(&typeString, &typeStringLen));
+
+            StackSString typeStringUtf8{ SString::Utf8, typeString, typeStringLen };
+
+            // Pass the string in the attribute to similar logic as Type.GetType(String).
+            // The below API will also handle any dependency between the returned type and the
+            // requesting assembly for the purposes of lifetime tracking of collectible types.
+            TypeHandle targetType = TypeName::GetTypeReferencedByCustomAttribute(
+                typeStringUtf8.GetUnicode(),
+                cxt.Declaration->GetAssembly(),
+                cxt.Declaration /* unsafeAccessorMethod */);
+            _ASSERTE(!targetType.IsNull());
+
+            // Future versions of the runtime may support
+            // UnsafeAccessorTypeAttribute on value types.
+            if (targetType.IsValueType())
+                ThrowHR(COR_E_NOTSUPPORTED, BFA_INVALID_UNSAFEACCESSORTYPE_VALUETYPE);
+
+            USHORT seq;
+            DWORD attr;
+            LPCSTR paramName;
+            IfFailThrow(pInternalImport->GetParamDefProps(currParamDef, &seq, &attr, &paramName));
+
+            if (seq >= totalParamCount)
+                ThrowHR(COR_E_BADIMAGEFORMAT, BFA_INVALID_UNSAFEACCESSORTYPE);
+
+            // UnsafeAccessorAttribute requires the return type to be byref for any field kind.
+            // This creates a type safety issue for the runtime. Since byrefs don't support
+            // contravariance, we can't permit returning a byref to a fully typed field as a
```

Comment:
```suggestion
            // covariance, we can't permit returning a byref to a fully typed field as a
```

Oops, wrong direction.